### ISSUE TITLE
pivot splitting the columns instead of grouping. 

### ISSUE THREAD

**Issue by AaronCreighton** [pola-rs/polars#12407]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

has it already been reported:
This issue is an related to the past issues found here: https://github.com/pola-rs/polars/issues/11663. This previous issue is still open and may need to be one first. 

This issue maybe related to https://github.com/pola-rs/polars/issues/12291. 

tested on latest version > 0.19.13 

The following is one example of the issue, there are likely a variety of input/outputs that have similar issues:
```python
df = pl.DataFrame({'a':['beep', 'bop'],'b':['a','b'],'c':['s','f'],'d':[7,8], 'e':['x', 'y']})
df.pivot(values=['a'], index='b', columns=['c', 'e'], aggregate_function='count')
```
shape: (2, 5)

| b   | s    | f    | x    | y    |
|-----|------|------|------|------|
| str | u32  | u32  | u32  | u32  |
| "a" | 1    | null | 1    | null |
| "b" | null | 1    | null | 1    |

:)

### Issue description

here is another example:

`df = pl.DataFrame({'a':['beep',None,'b'],'b':['a','b','b'],'c':['s','f','x'],'d':[7,8,9], 'e':['x', 'x','y']}) 
df.pivot(values=['a'], index='b', columns=['c', 'e'], aggregate_function='count')`

### Expected behavior

As with the previous issue https://github.com/pola-rs/polars/issues/11663, expected format would be:

| b   | s_x   | f _y   | 
|-----|------|------|
| str | u32  | u32  | 
| "a" | 1    | null | 
| "b" | null | 1    |

### Installed versions

<details>

```
Polars:              0.19.13
Index type:          UInt32
Platform:            Linux-5.4.0-1075-aws-x86_64-with-glibc2.27
Python:              3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.10.0
gevent:              <not installed>
matplotlib:          <not installed>
numpy:               1.26.0
openpyxl:            <not installed>
pandas:              2.1.1
pyarrow:             11.0.0
pydantic:            <not installed>
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          2.0.21
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

**Comment by cmdlineluser** [comment]
I've not used `.pivot` too much so this may be a silly question:

Is the problem that it is expected to behave like pandas `.pivot_table()` (but doesn't?)
```python
df = pl.DataFrame({
    'a':['beep',None,'b'],'b':['a','b','b'],'c':['s','f','x'],'d':[7,8,9], 'e':['x', 'x','y']
})

df.pivot(values=['a'], index='b', columns=['c', 'e'], aggregate_function='count')

# shape: (2, 6)
# ┌─────┬──────┬──────┬──────┬─────┬──────┐
# │ b   ┆ s    ┆ f    ┆ x    ┆ x   ┆ y    │
# │ --- ┆ ---  ┆ ---  ┆ ---  ┆ --- ┆ ---  │
# │ str ┆ u32  ┆ u32  ┆ u32  ┆ u32 ┆ u32  │
# ╞═════╪══════╪══════╪══════╪═════╪══════╡
# │ a   ┆ 1    ┆ null ┆ null ┆ 1   ┆ null │
# │ b   ┆ null ┆ 1    ┆ 1    ┆ 1   ┆ 1    │
# └─────┴──────┴──────┴──────┴─────┴──────┘
```
```python
pl.from_pandas(
    df.to_pandas().pivot_table(values=['a'], index='b', columns=['c', 'e'], aggfunc='count')
      .reset_index()
)

# shape: (2, 4)
# ┌───────────────┬─────────────────┬─────────────────┬─────────────────┐
# │ ('b', '', '') ┆ ('a', 'f', 'x') ┆ ('a', 's', 'x') ┆ ('a', 'x', 'y') │
# │ ---           ┆ ---             ┆ ---             ┆ ---             │
# │ str           ┆ f64             ┆ f64             ┆ f64             │
# ╞═══════════════╪═════════════════╪═════════════════╪═════════════════╡
# │ a             ┆ null            ┆ 1.0             ┆ null            │
# │ b             ┆ 0.0             ┆ null            ┆ 1.0             │
# └───────────────┴─────────────────┴─────────────────┴─────────────────┘
```

**Comment by AaronCreighton** [comment]
@cmdlineluser thank you for responding.

Pandas has pivot and pivot_table. 

Based on this slack post:
https://stackoverflow.com/questions/30960338/difference-between-pivot-and-pivot-table-why-is-only-pivot-table-working

- `pivot_table `is a generalization of pivot that can handle duplicate values for one pivoted index/column pair. Specifically, you can give pivot_table a list of aggregation functions using keyword argument aggfunc. The default aggfunc of pivot_table is numpy.mean.

- `pivot_table `also supports using multiple columns for the index and column of the pivoted table. A hierarchical index will be automatically generated for you.

So, a key question for polars Team is; Do we want to do the same? or have pivot do both? 

currently, based on the the flexibility of `polar.pivot()` it seems to align with `pandas.pivot_table()`.

Personally, I like one function. It means if I want to change what it does i can just update the one method. It also means that cheat sheets etc only have to have one function. e.g. with panda many cheat sheets will have just pivot.. then someone try something more complex with pivot and it won't work.. and have to find/dig up pivot_table via google.  In short, it makes it easy to expand the functionality with one method. 

I like two functions, because then replacing pandas code with polars becomes easier. 

from a pragmatic perspective: `polars.pivot()`  could continue along its current path (my perspective) and work like `pandas.pivot_table()` but then you could have `polars.pivot_table()` too, that just points to `polars.pivot()`.

**Comment by deanm0000** [comment]
This same issue came up back on 9/1 in the discord with no resolution...

@tikkanz had this:

```
df = pl.DataFrame([
    pl.Series("name", 4*["John", "Dick", "Mary", "Jane", "Bob"]),
    pl.Series("h", range(1,21)),
    pl.Series("w", [i/3 for i in range(1,21)]),
    pl.Series("dt", 2*((5*["m"]) + (5*["e"]))),
    pl.Series("mt", ((10*["a"]) + (10*["b"])))
]).sort(by=["name","dt", "mt"])
df.pivot(values=["h","w"], index="name", columns=["dt", "mt"], aggregate_function="first")
shape: (5, 9)
│ name ┆ h_dt_e ┆ h_dt_m ┆ w_dt_e   ┆ w_dt_m   ┆ h_mt_a ┆ h_mt_b ┆ w_mt_a   ┆ w_mt_b   │
│ ---  ┆ ---    ┆ ---    ┆ ---      ┆ ---      ┆ ---    ┆ ---    ┆ ---      ┆ ---      │
│ Bob  ┆ 10     ┆ 5      ┆ 3.333333 ┆ 1.666667 ┆ 10     ┆ 20     ┆ 3.333333 ┆ 6.666667 │
│ Dick ┆ 7      ┆ 2      ┆ 2.333333 ┆ 0.666667 ┆ 7      ┆ 17     ┆ 2.333333 ┆ 5.666667 │
│ Jane ┆ 9      ┆ 4      ┆ 3.0      ┆ 1.333333 ┆ 9      ┆ 19     ┆ 3.0      ┆ 6.333333 │
│ John ┆ 6      ┆ 1      ┆ 2.0      ┆ 0.333333 ┆ 6      ┆ 16     ┆ 2.0      ┆ 5.333333 │
│ Mary ┆ 8      ┆ 3      ┆ 2.666667 ┆ 1.0      ┆ 8      ┆ 18     ┆ 2.666667 ┆ 6.0      │
```

and was expecting

```
| name | h:dt=m,mt=a | h:dt=m,mt=b | h:dt=e,mt=a | h:dt=e,mt=b | w:dt=m,mt=a | w:dt=m,mt=b | w:dt=e,mt=a | w:dt=e,mt=b |
|------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
| Bob  | 10.0        | 20.0        | 5.0         | 15.0        | 3.333       | 6.667       | 1.667       | 5.000       |
| Dick | 7.0         | 17.0        | 2.0         | 12.0        | 2.333       | 5.667       | 0.667       | 4.000       |
| ..   |  ..         | ..          | ..          |  ..         | ..          | ..          |   ..        |  ..         |
```

And found this work around

```
df.with_columns(pl.concat_str(["dt","mt"], separator="_").alias("dt_mt")).pivot(
    values=["h","w"], index="name", columns="dt_mt", aggregate_function="first"
    )
```

There was some discussion [here](https://github.com/pola-rs/polars/pull/8002)


Back to this example, the same work around works like this

```
df.with_columns(ce=pl.concat_str(['c','e'], separator="_")).pivot(values=['a'], index='b', columns='ce', aggregate_function='count')
shape: (2, 3)
┌─────┬──────┬──────┐
│ b   ┆ s_x  ┆ f_y  │
│ --- ┆ ---  ┆ ---  │
│ str ┆ u32  ┆ u32  │
╞═════╪══════╪══════╡
│ a   ┆ 1    ┆ null │
│ b   ┆ null ┆ 1    │
└─────┴──────┴──────┘
```

@MarcoGorelli, ritchie tagged you back in Sep on this, any thoughts? Perhaps one way to do it is to have columns accept an expression so then it can be 

```
df.pivot(values=['a'], index='b', columns=pl.concat_str(['c','e'], separator="_"), aggregate_function='count')
```

The benefit here is that the there wouldn't be a fixed naming scheme and users can change it. Say you want the output to look like

```
shape: (2, 3)
┌─────┬───────────┬───────────┐
│ b   ┆ c=s|||e=x ┆ c=f|||e=y │
│ --- ┆ ---       ┆ ---       │
│ str ┆ u32       ┆ u32       │
╞═════╪═══════════╪═══════════╡
│ a   ┆ 1         ┆ null      │
│ b   ┆ null      ┆ 1         │
└─────┴───────────┴───────────┘
```

You could get that with 
`df.pivot(values=['a'], index='b', columns=pl.lit('c=') + pl.col('c') + pl.lit("|||") + pl.lit('e=')+pl.col('e'), aggregate_function='count')`

of course you can do that now with `df.with_columns(ce=pl.lit('c=') + pl.col('c') + pl.lit("|||") + pl.lit('e=')+pl.col('e')).pivot(values=['a'], index='b', columns='ce', aggregate_function='count')`

**Comment by MarcoGorelli** [comment]
hi @AaronCreighton 

I don't understand how you got your expected output, could you explain please?

You have
```python
In [107]: df
Out[107]:
shape: (3, 5)
┌──────┬─────┬─────┬─────┬─────┐
│ a    ┆ b   ┆ c   ┆ d   ┆ e   │
│ ---  ┆ --- ┆ --- ┆ --- ┆ --- │
│ str  ┆ str ┆ str ┆ i64 ┆ str │
╞══════╪═════╪═════╪═════╪═════╡
│ beep ┆ a   ┆ s   ┆ 7   ┆ x   │
│ null ┆ b   ┆ f   ┆ 8   ┆ x   │
│ b    ┆ b   ┆ x   ┆ 9   ┆ y   │
└──────┴─────┴─────┴─────┴─────┘
```
If you pivot with index `'b'`, then I'd expect:
- a column called `'b'`
- one row for each unique value in `'b'`

So far so good. Then, I'd expect a column for each unique combination of values in columns `'c'` and `'e'`. So:
- a column for `'s_x'`
- a column with `'f_x'`
- a column for `'x_y'`

I don't see how you got `'f_y'`

This would match the output from @deanm0000 's 
```python
(
    df.with_columns(
        ce=pl.concat_str("c", "e", separator="_")
    ).pivot(
        index="b",
        values="a",
        columns="ce",
        aggregate_function="count",
    )
)
```
```python
shape: (2, 4)
┌─────┬──────┬──────┬──────┐
│ b   ┆ s_x  ┆ f_x  ┆ x_y  │
│ --- ┆ ---  ┆ ---  ┆ ---  │
│ str ┆ u32  ┆ u32  ┆ u32  │
╞═════╪══════╪══════╪══════╡
│ a   ┆ 1    ┆ null ┆ null │
│ b   ┆ null ┆ 1    ┆ 1    │
└─────┴──────┴──────┴──────┘
```

If the suggestion is that this be the default output for `columns=['c', 'e']`, then I think that's a good suggestion

**Comment by AaronCreighton** [comment]
@MarcoGorelli Thank you for your patience, git notifications were being sent to the wrong folder (social i think)..

@deanm0000  that does looks close..

@MarcoGorelli, that last data frame 2,4 shape looks good. 

Here is the pandas version for comparison. 

`df = pd.DataFrame({'a':['beep',None,'b'],'b':['a','b','b'],'c':['s','f','x'],'d':[7,8,9], 'e':['x', 'x','y']}) 
df.pivot_table(index=['b'],columns=['c','e'], values='a', aggfunc='count') 
`

**Comment by cmdlineluser** [comment]
Looks like the pandas version doesn't count nulls?

```python
aggregate_function = pl.element().drop_nulls().count()
```

**Comment by AaronCreighton** [comment]
@cmdlineluser, i would not expect it to count nulls. I would expect a different aggfunc to count nulls something like countrows.

**Comment by cmdlineluser** [comment]
Aye. I was just showing how to get the same output as pandas.

```python
df.with_columns(ce=pl.concat_str("c", "e", separator="_")).pivot(index="b", values="a", columns="ce",
    aggregate_function = pl.count()
)        

# shape: (2, 4)
# ┌─────┬──────┬──────┬──────┐
# │ b   ┆ s_x  ┆ f_x  ┆ x_y  │
# │ --- ┆ ---  ┆ ---  ┆ ---  │
# │ str ┆ u32  ┆ u32  ┆ u32  │
# ╞═════╪══════╪══════╪══════╡
# │ a   ┆ 1    ┆ null ┆ null │
# │ b   ┆ null ┆ 1    ┆ 1    │ # <- f_x should be 0
# └─────┴──────┴──────┴──────┘
```

```python
df.with_columns(ce=pl.concat_str("c", "e", separator="_")).pivot(index="b", values="a", columns="ce",
   aggregate_function = pl.element().drop_nulls().count()
)        

# shape: (2, 4)
# ┌─────┬──────┬──────┬──────┐
# │ b   ┆ s_x  ┆ f_x  ┆ x_y  │
# │ --- ┆ ---  ┆ ---  ┆ ---  │
# │ str ┆ u32  ┆ u32  ┆ u32  │
# ╞═════╪══════╪══════╪══════╡
# │ a   ┆ 1    ┆ null ┆ null │
# │ b   ┆ null ┆ 0    ┆ 1    │
# └─────┴──────┴──────┴──────┘
```


Although, I thought count was recently changed so that it didn't count nulls.

**Comment by MarcoGorelli** [comment]
I think this can be tackled as part of https://github.com/pola-rs/polars/issues/12407

Closing in favour of that one then, thanks all for comments / report

**Comment by AaronCreighton** [comment]
does this mean I need to open a new issue?

I already asked in #12407 if I needed to create a new ticket, or if the issue was similar enough. They said, new ticket

**Comment by MarcoGorelli** [comment]
wrong link sorry, have updated

What I meant is - grouping the columns is the solution to https://github.com/pola-rs/polars/issues/11663, so this will be tackled as part of https://github.com/pola-rs/polars/issues/11663 and will hopefully be available in 1.0
